ext/standard/proc_open.c

```diff
@@ -323,15 +250,130 @@ static void proc_open_rsrc_dtor(zend_resource *rsrc)
 	_php_free_envp(proc->env);
 	efree(proc->pipes);
 	zend_string_release_ex(proc->command, false);
-	efree(proc);
+	proc->command = NULL;
+}
+
+static void php_process_free_obj(zend_object *object)
+{
+	php_process_handle *proc = php_process_from_obj(object);
+
+	if (proc->command == NULL) {
+		zend_object_std_dtor(&proc->std);
+		return;
+	}
+
+	php_process_free(proc);
+
+	zend_object_std_dtor(&proc->std);
+}
+
+static zend_function *php_process_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct Process, use proc_open() instead");
+	return NULL;
+}
+
+void php_register_process_class_handlers(void)
+{
+	process_ce->default_object_handlers = &process_object_handlers;
+	process_ce->create_object = php_process_create_object;
+
+	memcpy(&process_object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	process_object_handlers.offset = XtOffsetOf(php_process_handle, std);
+	process_object_handlers.free_obj = php_process_free_obj;
+	process_object_handlers.get_constructor = php_process_get_constructor;
+	process_object_handlers.clone_obj = NULL;
+	process_object_handlers.compare = zend_objects_not_comparable;
+}
+
+/* {{{ _php_array_to_envp
+ * Process the `environment` argument to `proc_open`
+ * Convert into data structures which can be passed to underlying OS APIs like `exec` on POSIX or
+ * `CreateProcessW` on Win32 */
+static php_process_env _php_array_to_envp(zval *environment)
+{
+	zval *element;
+	php_process_env env;
+	zend_string *key, *str;
+#ifndef PHP_WIN32
+	char **ep;
+#endif
+	char *p;
+	size_t sizeenv = 0;
+	HashTable *env_hash; /* temporary PHP array used as helper */
+
+	memset(&env, 0, sizeof(env));
+
+	if (!environment) {
+		return env;
+	}
+
+	uint32_t cnt = zend_hash_num_elements(Z_ARRVAL_P(environment));
+
+	if (cnt < 1) {
+#ifndef PHP_WIN32
+		env.envarray = (char **) ecalloc(1, sizeof(char *));
+#endif
+		env.envp = (char *) ecalloc(4, 1);
+		return env;
+	}
```

Comment:
I was confused as to why this was initialising stuff but not doing anything. Is this because ``cnt == 0`` effectively?

Why is this different to the no env case?